### PR TITLE
Caldav client

### LINKED ISSUE #20: CalDav-Interface: Create new events, Fetch data from CalDav server and make available in an internal representation

### PR DESCRIPTION
**Pull request by Skn0tt**
Closes #20

### CODE REVIEW

**Comment by phlprcks** [comment]
@Skn0tt @sspangenberg the coverage test fails, would it be possible to fix that before merging?

**Comment by phlprcks** [comment]
@Skn0tt and @sspangenberg are you working on this branch to fix the test

**Comment by sspangenberg** [comment]
> 
> 
> @Skn0tt and @sspangenberg are you working on this branch to fix the test?

We are investigating why the test works locally but fails in Github

File: app/caldav/index.ts [hunk 1]
```diff
@@ -26,18 +27,19 @@ async function makeRequestTo(
...
     },
     method: "REPORT",
-    data: `
-    <c:free-busy-query xmlns:c="urn:ietf:params:xml:ns:caldav">
-        <c:time-range start="20201023T000000Z" end="20201223T000000Z"/>
-    </c:free-busy-query>
-    `,
+    data: `<?xml version="1.0"?>
+           <c:free-busy-query xmlns:c="urn:ietf:params:xml:ns:caldav">
+           <c:time-range start="20201123T000000Z" end="20201127T000000Z" />
```
Comment (Skn0tt) [inline app/caldav/index.ts:39]: Can we have the time range be passed in via parameters?

Comment (sspangenberg) [reply app/caldav/index.ts:39]: Sure, I will change it when I add the other missing requests

File: app/caldav/index.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  return await urllib.request(calendar.url, {
+    ...(calendar.auth.digest ? { digestAuth: authString } : { auth: authString }),
+    method: method as any,
+    headers: headers as any,
+    data,
+  })
+}
+
+function formatDateString(date: Date) {
+  return date.toISOString().split('-').join("").split(':').join("").split('.')[0]+"Z"
```
Comment (Skn0tt) [inline app/caldav/index.ts:42]: I think `split("-").join("")` could be `replace("-", "")` instead

Comment (Skn0tt) [reply app/caldav/index.ts:42]: Also, I think we should add some unit tests for that function.

@sspangenberg have you checked the date formats for any edge-cases that aren't covered by this (presumably naive) algorithm?

File: app/caldav/index.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    data: `<?xml version="1.0"?>
+           <c:calendar-query xmlns:d="DAV:" xmlns:c="urn:ietf:params:xml:ns:caldav">
+           <d:prop>
+           <d:getetag />
+           <c:calendar-data />
+           </d:prop>
+           <c:filter>
+           <c:comp-filter name="VCALENDAR">
+   	       <c:comp-filter name="VEVENT">
+     		   <c:time-range start="${formattedStart}" end="${formattedEnd}">
```
Comment (Skn0tt) [inline app/caldav/index.ts:62]: I guess we should fix the indentation on this XML, just to make it kind of readable.

File: app/caldav/index.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+		       </c:time-range>
+	         </c:comp-filter>
+ 	         </c:comp-filter>
+           </c:filter>
+           </c:calendar-query>`,
+  })
+  const parsed = ical.sync.parseICS(response.data.toString())
```
Comment (Skn0tt) [inline app/caldav/index.ts:69]: Transforming ICS to our internal representation should go into it's own function.

File: app/caldav/index.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  for (let id of ids) {
+    if (parsed[id]["type"] === 'VEVENT') {
+      events.push({title: parsed[id]["summary"] as any, start: parsed[id]["start"] as any, end: parsed[id]["end"] as any})
+    }
+  }
+  events.sort((a, b) => (a.start > b.start) ? 1 : -1)
+  return events
+}
+
+// currently not working on default calender
```
Comment (Skn0tt) [inline app/caldav/index.ts:81]:
```suggestion
// currently not working on baikal's default calender
```

File: app/caldav/index.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    headers: {
+      Depth: 1,
+    },
+    method: "REPORT",
+    data: `<?xml version="1.0"?>
+           <c:free-busy-query xmlns:c="urn:ietf:params:xml:ns:caldav">
+           <c:time-range start="${formattedStart}" end="${formattedEnd}" />
+           </c:free-busy-query>`,
+  })
+  const parsed = ical.sync.parseICS(response.data.toString())
```
Comment (Skn0tt) [inline app/caldav/index.ts:95]: Should go into it's own function.

**Review by Skn0tt — COMMENTED** [review]
Looks mostly good to me.
